V4L2 Codec/camera fixes

6by9 opened the pull request:
@jc-kynesim Fixes for the kernel WARN due to us not returning any buffers that have been queued with mmal-vchiq before the port is enabled.

6by9 committed:
staging: vchiq-mmal: Add a function to return queued buffers

The mmal-vchiq layer allows submission of buffers before the port
is enabled, and it queues them to be sent on port enable.

There isn't a way of returning those buffers, which leads to
videobuf2 complaining of buffers not being returned by drivers.
Add such a call, and reuse the same code in port_disable.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging: bcm2835-camera: Ensure all queued buffers are returned

Buffers submitted before VIDIOC_STREAMON would be passed to the
mmal-vchiq layer, but weren't returned to videobuf2 if STREAMON
failed as there wasn't a call into mmal-vchiq to do so.

As there is now a suitable call, ensure buffers are returned.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging: bcm2835-codec: Ensure all buffers are returned on STREAMON failure

If STREAMON failed, then buffers weren't being returned to videobuf2,
leading to it producing WARNs in the kernel logs.

Return all buffers stored with the V4L2 M2M framework and with
mmal-vchiq should STREAMON fail.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging: bcm2835-camera: Disable camera on error

If the encode tunnel failed to enable, then the camera was left
enabled and that caused a failure on a subsequent request to
change format.

Disable the camera on failure.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c:
```diff
@@ -554,7 +554,8 @@ static int start_streaming(struct vb2_queue *vq, unsigned int count)
 						       flush_cb);
-			return -1;
+			disable_camera(dev);
+			return -EINVAL;
```
Is EINVAL the most appropriate error here? Is the return value from vchiq_mmal_port_enable not suitable?

6by9 replied:
> Is the return value from vchiq_mmal_port_enable not suitable?

Often not, although that's a different issue.

https://github.com/raspberrypi/linux/blob/rpi-5.10.y/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c#L1273 returns the MMAL error code without mapping it to Linux error codes, so you'll often see -3 [-MMAL_EINVAL](https://github.com/raspberrypi/userland/blob/master/interface/mmal/mmal_types.h#L51) appearing as [-ESRCH](https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/errno-base.h#L7)

6by9 replied:
OK, fixed up the error handling and returning the error from vchiq_mmal_port_enable both here and at line 599.

6by9 committed:
staging: mmal-vchiq: Clean up a load of !=0 and == 0 conditionals

Replace several instances of if (foo == 0) with if (!foo) and
if (foo != 0) with if (foo)

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging: mmal-vchiq: Map MMAL errors to Linux ones

MMAL error values were being returned to userspace by some of
the functions. Map them to appropriate Linux ones.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Hold fire - messed up in last changes.

6by9 commented:
OK, that should work now.

pelwell commented:
I'll wait for some feedback from @jc-kynesim, but LGTM.

timemaster5 commented:
Ok, I have tested this code, and the result is no HW decoding for video files. There are no errors reported, but I see only a black square instead of a rendered picture. Don't you know whether a newer GStreamer is required for this to work? I am on 1.16.

6by9 commented:
@timemaster5 GStreamer issue may well be the same as https://github.com/raspberrypi/firmware/issues/1673

I'll pull these patches forward to 5.15 as 5.10 is effectively frozen now.
